mmm_experiments/agents/scientific_value.py

```diff
@@ -0,0 +1,252 @@
+
+    return v.mean(axis=1)
+
+
+class ScientificValueAgentMixin:
+    """This is a jack-of-all-trades agent which is general to any type of
+    observation (i.e. any beamline). This is a _parameter free_ method, meaning
+    that the user does not need to set any parameters explicitly, though one
+    can manually set a length scale if desired.
+
+    Assumes that ``device``, ``relative_bounds`` and ``measurement_origin`` are
+    defined as attributes.
+    """
+
+    def __init__(
+        self,
+        length_scale: Optional[float] = None,
+        y_distance_function: Optional[callable] = None,
+        optimize_acqf_kwargs: dict = {
+            "q": 1, "num_restarts": 5, "raw_samples": 20
+        },
+        possible_coordinates: Optional[np.ndarray] = None,
+        **kwargs
+    ):
+        super().__init__(**kwargs)
+        self._length_scale = length_scale
+        self._y_distance_function = y_distance_function
+        self._optimize_acqf_kwargs = optimize_acqf_kwargs
+
+        # Possible coordinatse should be an array of shape L x d, where L is
+        # the number of possible coordinates and d is the dimension of the
+        # space (i.e. it's 2 if we have a 2-dimensional scanning surface)
+        self._possible_coordinates = possible_coordinates
+
+        # Caches for the results
+        self._observations_cache = []
+        self._positions_cache = []
+        self._relative_positions_cache = []
+        self._value_cache = []
+
+    def tell(self, position, observation):
+        """Takes the position of the motor and an arbitrary observation which
+        is a function of that position, computes the value of this site, and
+        appends the proper caches.
+
+        Parameters
+        ----------
+        position : array_like
+            The position of the motor. Must match the dimensionality of that
+            provided in the ``__init__``.
+        observation : array_like
+            The observation, such as a spectrum, PDF, XRD, etc. Must match the
+            dimensionality of that provided in the ``__init__``.
+
+        Returns
+        -------
+        dict
+        """
+
+        # Using the relative position for all learning
+        relative_position = position - self.measurement_origin
+
+        self._positions_cache.append(position)
+        self._relative_positions_cache.append(relative_position)
+        self._observations_cache.append(observation)
+
+        # Compute the value
+        X = np.array(self._relative_positions_cache)
+        Y = np.array(self._observations_cache)
+        V = scientific_value_function(X, Y, sd=self._length_scale)
+
+        # The value is a scalar
+        V = V.reshape(-1, 1)
+
+        return dict(
+            position=[position],
+            rel_position=[relative_position],
+            observation=observation,
+            cache_len=[len(self._relative_positions_cache)],
+            value=V.squeeze()[-1]
+        )
```

Comment:
```suggestion
        return dict(
            position=[position],
            rel_position=[relative_position],
            observation=[observation],
            cache_len=[len(self._relative_positions_cache)],
            value=[V.squeeze()[-1]]
        )
```
See below, listify everything because...well...reasons...